### PR TITLE
Add external authentication mechanism

### PR DESCRIPTION
**Pull request by fredipevcin**

### CODE REVIEW

**Comment by Neustradamus** [comment]
@fredipevcin: Why have you closed your PR?

**Comment by fredipevcin** [comment]
@Neustradamus: because there were no activity.

**Comment by Neustradamus** [comment]
@mattn: What do you think about this PR?

**Comment by mattn** [comment]
If you

**Comment by mattn** [comment]
Could you please fix conflict?

**Comment by Neustradamus** [comment]
@fredipevcin: Have you seen the @mattn comment?

**Comment by mattn** [comment]
Could you please fix conflict?

**Comment by fredipevcin** [comment]
Hey, I will try to fix this as soon as possible (week or 2). You are welcome to do it otherwise I will when time allows me.

**Comment by fredipevcin** [comment]
Had an issue with tests.. so I removed them. If you have a suggestion how to add them, let me know.

Another thing I noticed during the testing. If server respond with multiple mechanism then the first one will be selected (`PLAIN`). Not sure if this is good thing.
```xml
<stream:stream xmlns='jabber:client'
	   xmlns:stream='http://etherx.jabber.org/streams'
	   to='domain'
	   version='1.0'>
	<stream:features>
		<mechanisms xmlns='urn:ietf:params:xml:ns:xmpp-sasl'>
			<mechanism>PLAIN</mechanism>
			<mechanism>X-OAUTH2</mechanism>
		</mechanisms>
	</stream:features>
```

**Comment by fredipevcin** [comment]
@Neustradamus @mattn If this is good enough for you you can merge it. It's been a long time since I use this. I can't even test it.

**Comment by mdosch** [comment]
It should prefer SCRAM mechanisms over PLAIN. If you want to use OAUTH or EXTERNAL I guess you'll tell the lib to do so by the setting.

I have no idea about tests. Maybe @mattn knows more.

Am 24. November 2023 12:59:01 UTC schrieb Fredi Pevcin ***@***.***>:
>Had an issue with tests.. so I removed them. If you have a suggestion how to add them, let me know.
>
>Another thing I noticed during the testing. If server respond with multiple mechanism then the first one will be selected (`PLAIN`). Not sure if this is good thing.
>```xml
><stream:stream xmlns='jabber:client'
>	   xmlns:stream='http://etherx.jabber.org/streams'
>	   to='domain'
>	   version='1.0'>
>	<stream:features>
>		<mechanisms xmlns='urn:ietf:params:xml:ns:xmpp-sasl'>
>			<mechanism>PLAIN</mechanism>
>			<mechanism>X-OAUTH2</mechanism>
>		</mechanisms>
>	</stream:features>
>```
>
>
>

**Comment by Neustradamus** [comment]
@fredipevcin: Have you seen the @mdosch comment?

**Comment by Neustradamus** [comment]
@fredipevcin: What is your email address?

**Comment by mdosch** [comment]
@fredipevcin I also don't have a server using external authentication and am therfore also unable to test it. :(

File: xmpp.go [hunk 1]
```diff
@@ -181,6 +181,9 @@ type Options struct {
 	// provided to the server as the xmlns:auth attribute of the OAuth2 authentication request.
 	OAuthXmlNs string
 
+	// AuthExternal will be chosen if this is set to true and if server supports it
+	AuthExternal bool
```
Comment (mdosch) [inline xmpp.go:185]: I think `Mechanism` could also be used for setting auth to `EXTERNAL`: https://github.com/xmppo/go-xmpp/blob/master/xmpp.go#L234

**Comment by Neustradamus** [comment]
@fredipevcin: Have you seen the @mdosch comments?

**Comment by fredipevcin** [comment]
@Neustradamus I've seen. I suggest to close this PR because it's 6 years old and I'will not resolve conflicts again. If someone has time to implement this, please do.

**Pull request closed without merging**